### PR TITLE
feat: Add Carthage support

### PR DESCRIPTION
**Pull request by ntadej**
## Changes

This adds manager support for [Carthage](https://github.com/Carthage/Carthage). To make it work an additional versioning scheme and data source have also been added. The code is based on existing codebase and I am open to suggestions if I copied legacy/poor examples.

## Context

Carthage is one of the iOS/macOS dependency managers that is relatively simple and standalone. It has a package file and a lock file that use simplified semver with only `==`, `>=` and `~>` operators supported. The data sources are either git repositories or simple JSON files published by dependency authors.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Test repository: https://github.com/ntadej/test-renovate/pulls

### CODE REVIEW

File: lib/modules/manager/carthage/extract.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import type { PackageDependency, PackageFileContent } from '../types';
+import type { ManagerData, ParsedLine } from './types';
+
+const regexMappings = [
+  regEx(
+    `^\\s*(?<type>[^\\s]+)\\s+(['"])(?<url>[^'"]+)(['"])\\s*(['"])(?<version>[^'"]+)(['"])\\s*$`
+  ),
+  regEx(
+    `^\\s*(?<type>[^\\s]+)\\s+(['"])(?<url>[^'"]+)(['"])\\s*(?<condition>>=|==|~>)\\s*(?<version>[^\\s]+)\\s*$`
+  ),
```
Comment (secustor) [inline lib/modules/manager/carthage/extract.ts:18]: You shouldn't need to decide this here, but only return `currentValue` with `== 1.0.0` or `>= 1.0.0`. The analysis of the `currentValue` is the job of the versioning

This allows you to simplify to one regex

File: lib/modules/manager/carthage/types.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export interface ParsedLine {
+  url?: string;
+  type?: string;
```
Comment (secustor) [inline lib/modules/manager/carthage/types.ts:3]:
```suggestion
  url: string;
  type: string;
```
Shouldn't `url` and `type` always present?

File: lib/modules/manager/carthage/extract.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        depName: parseDepName(url),
+        skipReason: 'unspecified-version',
+      };
+
+      if (url.startsWith('file://') || url.startsWith('/')) {
+        dep = {
+          depName: parseDepName(url),
+          skipReason: 'local-dependency',
+        };
+      } else if (type === 'github' || type === 'git') {
```
Comment (secustor) [inline lib/modules/manager/carthage/extract.ts:105]: Shouldn't this also include `gitlab`?

File: lib/modules/manager/carthage/extract.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    const { type, url, version, condition }: ParsedLine = parsedLine;
+
+    if (url) {
+      const managerData: ManagerData = { lineNumber };
+      let dep: PackageDependency = {
+        depName: parseDepName(url),
+        skipReason: 'unspecified-version',
+      };
+
+      if (url.startsWith('file://') || url.startsWith('/')) {
```
Comment (secustor) [inline lib/modules/manager/carthage/extract.ts:100]: Can this be local paths too? `./`

File: lib/modules/manager/carthage/extract.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        dep = {
+          depName: parseDepName(url),
+          skipReason: 'local-dependency',
+        };
+      } else if (type === 'github' || type === 'git') {
+        if (version && condition) {
+          dep = { ...gitDep(parsedLine), managerData };
+        } else {
+          if (url.startsWith('http') || version) {
```
Comment (secustor) [inline lib/modules/manager/carthage/extract.ts:109]: I don't understand this condition. Why does this which skipReason is used?

File: lib/modules/manager/carthage/extract.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  } else if (platformMatch?.groups) {
+    const { account, repo, platform } = platformMatch.groups;
+    if (account && repo) {
+      const datasource =
+        platform === 'github'
+          ? GithubTagsDatasource.id
+          : GitlabTagsDatasource.id;
+      return {
+        datasource,
+        depName: parseDepName(url),
```
Comment (secustor) [inline lib/modules/manager/carthage/extract.ts:65]: Do not add `depName`, if the value is an empty string

**Review by secustor — CHANGES_REQUESTED** [review]
Thanks for your contribution. 

Be aware as there has been no implementation discussion the review process can take longer. 

This is my first pass.

File: lib/modules/manager/carthage/readme.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+The `carthage` manager supports extracting both binary distributions as well as dependencies on GitHub repositories that use versioned tags.
+Unversioned dependencies and dependencies targeting a branch are not supported.
+
+Whenever a `Cartfile` is updated, Renovate will also update any associated `Cartfile.resolved` file.
+This is done manually without running `carthage` by just replacing the raw version which should be valid in most of the cases.
```
[the remark this answers is not in the document]
Comment (ntadej) [reply lib/modules/manager/carthage/readme.md:5]: Replying to this one now: Will improve the text - I'm just not sure what happens with pre-release versions. I guess those would anyways be skipped.

File: lib/modules/manager/carthage/extract.ts (hunk 3 above)
Comment (ntadej) [reply lib/modules/manager/carthage/extract.ts:105]: Carthage does not explicitly support `gitlab` but the full URL can be parsed and falls under `git`.

Comment (secustor) [reply lib/modules/manager/carthage/extract.ts:105]: In that case use 
https://github.com/renovatebot/renovate/blob/f70f0c82beda4712f11f93a4a111ca4f56d4ea5a/lib/util/common.ts#L16

This takes Renovate configuration also into account, when detecting platforms.

**Comment by ntadej** [comment]
Thanks for the already detailed review. I know I should have probably discussed it with the community in advance, but quick prototyping resolved in fully working Carthage support. I'm travelling this week but will get back to it when I'm back and I'm happy to iterate on the implementation.

**Review by viceice — COMMENTED** [review]
please create separate PR's for versioning and datasource, so the PR's are smaller and faster to review

**Comment by rarkins** [comment]
Hi @ntadej do you have time to resume this work?

**Pull request closed without merging**